expression/builtin_time.go

```diff
@@ -2344,159 +2378,394 @@ func isDuration(str string) bool {
 	if err != nil {
-		return result, errors.Trace(err)
+		return "", errors.Trace(err)
 	}
 	tmpDuration, err := arg0.Add(arg1)
 	if err != nil {
-		return result, errors.Trace(err)
+		return "", errors.Trace(err)
 	}
-	if getFsp(d) != 0 {
-		tmpDuration.Fsp = types.MaxFsp
-	} else {
+	tmpDuration.Fsp = types.MaxFsp
+	if tmpDuration.MicroSecond() == 0 {
 		tmpDuration.Fsp = types.MinFsp
 	}
-	result.SetString(tmpDuration.String())
-	return
+	return tmpDuration.String(), nil
 }
 
-// strDatetimeSubDuration subtracts duration from datetime string, returns a datum value.
-func strDatetimeSubDuration(d string, arg1 types.Duration) (result types.Datum, err error) {
-	arg0, err := types.ParseTime(d, mysql.TypeDatetime, getFsp(d))
+// strDatetimeSubDuration subtracts duration from datetime string, returns a string value.
+func strDatetimeSubDuration(d string, arg1 types.Duration) (string, error) {
+	arg0, err := types.ParseTime(d, mysql.TypeDatetime, types.MaxFsp)
 	if err != nil {
-		return result, errors.Trace(err)
+		return "", errors.Trace(err)
 	}
 	arg1time, err := arg1.ConvertToTime(uint8(getFsp(arg1.String())))
 	if err != nil {
-		return result, errors.Trace(err)
+		return "", errors.Trace(err)
 	}
 	tmpDuration := arg0.Sub(&arg1time)
+	fsp := types.MaxFsp
+	if tmpDuration.MicroSecond() == 0 {
+		fsp = types.MinFsp
+	}
 	resultDuration, err := tmpDuration.ConvertToTime(mysql.TypeDatetime)
 	if err != nil {
-		return result, errors.Trace(err)
+		return "", errors.Trace(err)
 	}
-	if getFsp(d) != 0 {
-		tmpDuration.Fsp = types.MaxFsp
-	} else {
-		tmpDuration.Fsp = types.MinFsp
-	}
-	result.SetString(resultDuration.String())
-	return
+	resultDuration.Fsp = fsp
+	return resultDuration.String(), nil
 }
 
-// strDurationSubDuration subtracts duration from duration string, returns a datum value.
-func strDurationSubDuration(d string, arg1 types.Duration) (result types.Datum, err error) {
-	arg0, err := types.ParseDuration(d, getFsp(d))
+// strDurationSubDuration subtracts duration from duration string, returns a string value.
+func strDurationSubDuration(d string, arg1 types.Duration) (string, error) {
+	arg0, err := types.ParseDuration(d, types.MaxFsp)
 	if err != nil {
-		return result, errors.Trace(err)
+		return "", errors.Trace(err)
 	}
 	tmpDuration, err := arg0.Sub(arg1)
 	if err != nil {
-		return result, errors.Trace(err)
+		return "", errors.Trace(err)
 	}
-	if getFsp(d) != 0 {
-		tmpDuration.Fsp = types.MaxFsp
-	} else {
+	tmpDuration.Fsp = types.MaxFsp
+	if tmpDuration.MicroSecond() == 0 {
 		tmpDuration.Fsp = types.MinFsp
 	}
-	result.SetString(tmpDuration.String())
-	return
+	return tmpDuration.String(), nil
 }
 
 type addTimeFunctionClass struct {
 	baseFunctionClass
 }
 
-func (c *addTimeFunctionClass) getFunction(ctx context.Context, args []Expression) (builtinFunc, error) {
-	if err := c.verifyArgs(args); err != nil {
+func (c *addTimeFunctionClass) getFunction(ctx context.Context, args []Expression) (sig builtinFunc, err error) {
```

Comment:
extract a common to implement the common `getFunction` behavior?

Reply:
OK